exec/java-exec/src/main/java/org/apache/drill/exec/store/pcap/PcapBatchReader.java

```diff
@@ -0,0 +1,323 @@
+        // shift data and read more. This is the common case.
+        System.arraycopy(buffer, offset, buffer, 0, validBytes - offset);
+        validBytes = validBytes - offset;
+        offset = 0;
+
+        int n = fsStream.read(buffer, validBytes, buffer.length - validBytes);
+        if (n > 0) {
+          validBytes += n;
+        }
+        logger.info("read {} bytes, at {} offset", n, validBytes);
+      } else {
+        // near the end of the file, we will just top up the buffer without shifting
+        int n = fsStream.read(buffer, offset, buffer.length - offset);
+        if (n > 0) {
+          validBytes = validBytes + n;
+          logger.info("Topped up buffer with {} bytes to yield {}\n", n, validBytes);
+        }
+      }
+    } catch (Exception e) {
+      // Exception denotes EOF or unreadable file
+      return false;
+    }
+    return true;
+  }
+
+  private boolean addDataToTable(Packet packet, int networkType, RowSetLoader rowWriter) {
+    rowWriter.start();
+    String data;
+
+    typeWriter.setString(packet.getPacketType());
+    timestampWriter.setTimestamp(new Instant(packet.getTimestamp()));
+    timestampMicroWriter.setLong(packet.getTimestampMicro());
+    networkWriter.setInt(networkType);
+    srcMacAddressWriter.setString(packet.getEthernetSource());
+    dstMacAddressWriter.setString(packet.getEthernetDestination());
+
+    dstIPWriter.setString(packet.getDst_ip().getHostAddress());
+    srcIPWriter.setString(packet.getSrc_ip().getHostAddress());
+    srcPortWriter.setInt(packet.getSrc_port());
+    dstPortWriter.setInt(packet.getDst_port());
+    packetLengthWriter.setInt(packet.getPacketLength());
+
+    // TCP Only Packet Data
+    tcpSessionWriter.setLong(packet.getSessionHash());
+    tcpSequenceWriter.setInt(packet.getSequenceNumber());
+    tcpAckNumberWriter.setInt(packet.getAckNumber());
+    tcpFlagsWriter.setInt(packet.getFlags());
+    tcpParsedFlagsWriter.setString(packet.getParsedFlags());
+
+    // TCP Flags
+    tcpNsWriter.setBoolean((packet.getFlags() & 0x100) != 0);
+    tcpCwrWriter.setBoolean((packet.getFlags() & 0x80) != 0);
+    tcpEceWriter.setBoolean((packet.getFlags() & 0x40) != 0);
+    tcpFlagsEceEcnCapableWriter.setBoolean((packet.getFlags() & 0x42) == 0x42);
+    tcpFlagsCongestionWriter.setBoolean((packet.getFlags() & 0x42) == 0x40);
+    tcpUrgWriter.setBoolean((packet.getFlags() & 0x20) != 0);
+    tcpAckWriter.setBoolean((packet.getFlags() & 0x10) != 0);
+    tcpPshWriter.setBoolean((packet.getFlags() & 0x8) != 0);
+    tcpRstWriter.setBoolean((packet.getFlags() & 0x4) != 0);
+    tcpSynWriter.setBoolean((packet.getFlags() & 0x2) != 0);
+    tcpFinWriter.setBoolean((packet.getFlags() & 0x1) != 0);
+
+    // Note:  getData() MUST be called before isCorrupt
+    dataWriter.setString(parseBytesToASCII(packet.getData()));
+    isCorruptWriter.setBoolean(packet.isCorrupt());
+    rowWriter.save();
+
+    // TODO Parse Data Packet Here
```

Comment:
What this todo means?

Reply:
Well, we have a need to leave todo-s in production code they must contain enough information for other to understand. For example, link to Jira or extended description.